### PR TITLE
refactor: extract a function to create bindings of (non-)prefabs

### PR DESCRIPTION
**Pull request by capfredf**

### CODE REVIEW

File: typed-racket-lib/typed-racket/typecheck/tc-structs.rkt [hunk 1]
```diff
@@ -234,158 +238,118 @@
...
+                                    (->* (list poly-base) self-fld)
+                                    (->acc (list poly-base) self-fld (list path))))])
+                    (values (list poly-base idx-parent-cnt mutable #f)
+                            func)))
+                (lambda (_ self-fld idx-parent-cnt poly-base)
+                  (values (list poly-base idx-parent-cnt #f)
+                          (poly-wrapper (->* (list poly-base self-fld) -Void))))))
+
+
+(define-syntax-rule (nullable condition expr)
```
Comment (samth) [inline typed-racket-lib/typed-racket/typecheck/tc-structs.rkt:311]: Better than this abstraction is a `maybe-cons` function.

File: typed-racket-lib/typed-racket/typecheck/tc-structs.rkt [hunk 2]
```diff
@@ -234,158 +238,118 @@
...
+    names)
+  
+  (define all-fields (struct-desc-all-fields desc))
+  (define parent-count (struct-desc-parent-count desc))
+
+  ;; the base-type, with free type variables
+  (define poly-base
+    (mk-poly-base type-name tvars))
+
+  (define (mk-operator-bindings! operators add-struct-operator-fn! mk-vals)
```
Comment (samth) [inline typed-racket-lib/typed-racket/typecheck/tc-structs.rkt:335]: Signature/purpose

File: typed-racket-lib/typed-racket/typecheck/tc-structs.rkt [hunk 3]
```diff
@@ -212,19 +212,23 @@
     [(? Struct?) (register-non-prefab-bindings! sty names desc si)]
     [(? Prefab?) (register-prefab-bindings! sty names desc si)]))
 
+(define (mk-poly-base type-name tvars)
```
Comment (samth) [inline typed-racket-lib/typed-racket/typecheck/tc-structs.rkt:215]: signature/purpose/contract?

File: typed-racket-lib/typed-racket/typecheck/tc-structs.rkt [hunk 4]
```diff
@@ -234,158 +238,118 @@
...
+                  (values (list poly-base idx-parent-cnt #f)
+                          (poly-wrapper (->* (list poly-base self-fld) -Void))))))
+
+
+(define-syntax-rule (nullable condition expr)
+  (if condition
+      expr
+      null))
+
+(define (mk-bindings! sty si names desc mk-pred-ty mk-getter-vals mk-setter-vals)
```
Comment (samth) [inline typed-racket-lib/typed-racket/typecheck/tc-structs.rkt:316]: signature/purpose/contract

this seems like the important one, so explaining this well is what makes the refactoring worth it.

File: typed-racket-lib/typed-racket/typecheck/tc-structs.rkt [hunk 5]
```diff
@@ -348,44 +289,110 @@
...
+                                (if mutable
+                                    (->* (list poly-base) self-fld)
+                                    (->acc (list poly-base) self-fld (list path))))])
+                    (values (list poly-base idx-parent-cnt mutable #f)
+                            func)))
+                (lambda (_ self-fld idx-parent-cnt poly-base)
+                  (values (list poly-base idx-parent-cnt #f)
+                          (poly-wrapper (->* (list poly-base self-fld) -Void))))))
+
+;; if the condition, cons the value of the thunk onto the list
```
[the remark this answers is not in the document]
Comment (capfredf) [reply typed-racket-lib/typed-racket/typecheck/tc-structs.rkt:316]: Option A is better indeed. But in this context, I think the last argument to maybe-cons should be `(-> (Listof A))`,  especially for https://github.com/racket/typed-racket/pull/1091/files#diff-516cab1093154106286b0cff961a6e54c2468f64a452f70f4bb8a47b73e82df3R389

File: typed-racket-lib/typed-racket/typecheck/tc-structs.rkt [hunk 6]
```diff
@@ -348,44 +277,100 @@
...
+                         [func (poly-wrapper
+                                (if mutable
+                                    (->* (list st-type-alias) self-fld)
+                                    (->acc (list st-type-alias) self-fld (list path))))])
+                    (values (list st-type-alias idx-parent-cnt mutable #f) func)))
+                (lambda (_ self-fld idx-parent-cnt st-type-alias)
+                  (values (list st-type-alias idx-parent-cnt #f) (poly-wrapper (->* (list st-type-alias self-fld) -Void))))))
+
+;; if val-th is not false, evaluate it and cons the result onto rst
+;; (-> (Option (-> A) (Listof A) (Listof A))
```
Comment (samth) [inline typed-racket-lib/typed-racket/typecheck/tc-structs.rkt:302]: I think there are some paren issues here.

Comment (samth) [inline typed-racket-lib/typed-racket/typecheck/tc-structs.rkt:302]: I don't understand why the thunking is useful here.

File: typed-racket-lib/typed-racket/typecheck/tc-structs.rkt [hunk 7]
```diff
@@ -212,134 +211,64 @@
...
-    bindings)))
+  (mk-bindings! pty names desc si
+                (lambda (_) prefab-top-type)
+                (lambda (acc-id t idx _)
+                  (define path (make-PrefabPE key idx))
+                  (define fld-sym (list-ref field-tvar-syms idx))
+                  (define fld-t (list-ref field-tvar-Fs idx))
+                  (define func-t (cond
+                                   [(or self-mutable parent-mutable)
+                                    ;; NOTE - if we ever track mutable fields more ganularly
```
Comment (samth) [inline typed-racket-lib/typed-racket/typecheck/tc-structs.rkt:250]:
```suggestion
                                    ;; NOTE - if we ever track mutable fields more granularly
```

File: typed-racket-lib/typed-racket/typecheck/tc-structs.rkt (hunk 6 above)
Comment (capfredf) [reply typed-racket-lib/typed-racket/typecheck/tc-structs.rkt:302]: https://github.com/racket/typed-racket/pull/1091/files#diff-516cab1093154106286b0cff961a6e54c2468f64a452f70f4bb8a47b73e82df3R369

In this case, semantically, if mutators don't exist at all, we don't need to compute an empty list?

Comment (capfredf) [reply typed-racket-lib/typed-racket/typecheck/tc-structs.rkt:302]: You are right. Thunking is unnecessary

**Review by samth — APPROVED** [review]

**Pull request merged**